resharper/resharper-unity/src/CSharp/Feature/Services/QuickFixes/MoveQuickFixes/AbstractMoveQuickFix.cs

```diff
@@ -0,0 +1,151 @@
+using System;
+using System.Collections.Generic;
+using JetBrains.Annotations;
+using JetBrains.Application.Progress;
+using JetBrains.ProjectModel;
+using JetBrains.ReSharper.Feature.Services.Bulbs;
+using JetBrains.ReSharper.Feature.Services.Intentions;
+using JetBrains.ReSharper.Feature.Services.QuickFixes;
+using JetBrains.ReSharper.Feature.Services.Resources;
+using JetBrains.ReSharper.Plugins.Unity.Utils;
+using JetBrains.ReSharper.Psi.CSharp;
+using JetBrains.ReSharper.Psi.CSharp.Tree;
+using JetBrains.ReSharper.Psi.CSharp.Util;
+using JetBrains.ReSharper.Psi.Naming.Settings;
+using JetBrains.ReSharper.Psi.Tree;
+using JetBrains.TextControl;
+using JetBrains.Util;
+
+namespace JetBrains.ReSharper.Plugins.Unity.CSharp.Feature.Services.QuickFixes.MoveQuickFixes
+{
+    public abstract class AbstractMoveQuickFix : IQuickFix
+    {
+        [CanBeNull] protected readonly IClassDeclaration MonoBehaviourScript;
+        [CanBeNull] protected readonly ICSharpExpression ToMove;
+        [CanBeNull] protected readonly string FieldName;
+
+        public AbstractMoveQuickFix([CanBeNull] IClassDeclaration monoBehaviourScript, [CanBeNull]ICSharpExpression toMove, [CanBeNull] string fieldName = null)
+        {
+            MonoBehaviourScript = monoBehaviourScript;
+            ToMove = toMove;
+            FieldName = fieldName;
+        }
+        
+        public virtual IEnumerable<IntentionAction> CreateBulbItems()
+        {
+            Assertion.Assert(ToMove != null, "ToMove != null");
+            Assertion.Assert(MonoBehaviourScript != null, "MonoBehaviourScript != null");
+
+            var result = new List<IntentionAction>();
+            if (MonoBehaviourMoveUtil.IsExpressionAccessibleInScript(ToMove))
+            {
+                result.Add(new IntentionAction(new MoveAction(MonoBehaviourScript, ToMove, "Start", FieldName), BulbThemedIcons.ContextAction.Id, IntentionsAnchors.ContextActionsAnchor ));
+                result.Add(new IntentionAction(new MoveAction(MonoBehaviourScript, ToMove, "Awake", FieldName), BulbThemedIcons.ContextAction.Id, IntentionsAnchors.ContextActionsAnchor ));
+            }
+
+            var loopAction = TryCreateMoveFromLoopAction(ToMove);
+            if (loopAction != null)
+                result.Add(loopAction);
+
+            return result;
+        }
+
+        private IntentionAction TryCreateMoveFromLoopAction([NotNull] ICSharpExpression toMove)
+        {
+            ITreeNode currentExpression = toMove;
+            ILoopStatement previousLoop = null;
+
+            int closureStartOffset = currentExpression.GetContainingNode<ICSharpClosure>()?.GetTreeStartOffset().Offset ?? 0;
```

Comment:
I believe something like this will be shorter and more expressive than offset checks. It also does not require multiple traversals of containing nodes in order to find out if there is a closure among them at all.
```cs
foreach (var node in currentExpression.ContainingNodes())
{
  if (node is ICSharpClosure) break;
  if (node is ILoopStatement loop)
  {
    if (IsAvailableToMoveFromLoop())
      previousLoop = loop;
    else
     break;
  }
}
```